### PR TITLE
Adding section on derived array types and array subtypes

### PR DESCRIPTION
**Pull request by gusthoff**
Also: improving section on array constraints.

### CODE REVIEW

File: content/courses/advanced-ada/parts/data_types/arrays.rst [hunk 1]
```diff
@@ -52,21 +73,59 @@ also highlights the fact that the bounds are fixed once an object is declared:
...
+cannot specify the bounds (or change them) in the declaration of the :ada:`M`
+array, as they have already been defined in the type declaration.
 
 
 Unconstrained Arrays vs. Vectors
-~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 
 If you need, however, the flexibility of increasing the length of an array, you
 could use vectors instead. This is how we could rewrite the previous example
```
Comment (pat-rogers) [inline content/courses/advanced-ada/parts/data_types/arrays.rst:125]: Maybe "could use the language-defined Vector type instead."

File: content/courses/advanced-ada/parts/data_types/arrays.rst [hunk 2]
```diff
@@ -525,16 +585,182 @@ we'd like to highlight the following aspects:
...
+a range from 1 to 10.
+
+For the :ada:`Measurements_Max_20` subtype, we're specifying |mdash| via a
+dynamic predicate |mdash| that arrays of this subtype can only have 20
+components at most. Finally, for the :ada:`Measurements_First_Two` subtype,
+we're constraining the type to a range from 1 to 10 and requiring that the
+first component must have a value of 2.0.
+
+Note that we cannot set the default component value for array subtypes |mdash|
+only type declarations are allowed to use that.
```
Comment (pat-rogers) [inline content/courses/advanced-ada/parts/data_types/arrays.rst:734]: Either "are allowed to do that."
or
"are allowed to use that facility."
or something like that.

File: content/courses/advanced-ada/parts/data_types/arrays.rst (hunk 1 above)
Comment (gusthoff) [reply content/courses/advanced-ada/parts/data_types/arrays.rst]: Yes, that sounds better. Thanks for the suggestion!

File: content/courses/advanced-ada/parts/data_types/arrays.rst (hunk 2 above)
Comment (gusthoff) [reply content/courses/advanced-ada/parts/data_types/arrays.rst]: Indeed, thanks for the correction!

**Review by pat-rogers — APPROVED** [review]

**Pull request merged**